File: examples/01-basic-actor/go/counter.go
/**
 * Basic Actor Pattern - Counter Actor in Go with Proto.Actor
 *
 * 의존성 설치:
 * go get github.com/asynkron/protoactor-go
 */

package main

import (
	"fmt"
	"sync"
	"time"

	"github.com/asynkron/protoactor-go/actor"
)

// ============================================
// Message Types (메시지 타입 정의)
// ============================================

// Tell 메시지 (응답 없음)
type Increment struct{}
type Decrement struct{}
type Reset struct{}

// Ask 메시지 (응답 필요)
type GetCount struct{}
type CountResponse struct {
	Count int
}

// ============================================
// Counter Actor
// ============================================

type CounterActor struct {
	count int
}

func (c *CounterActor) Receive(ctx actor.Context) {
	switch msg := ctx.Message().(type) {
	case *actor.Started:
		fmt.Println("Counter Actor started")

	case *actor.Stopped:
		fmt.Println("Counter Actor stopped")

	case *Increment:
		c.count++
		fmt.Printf("Incremented: %d\n", c.count)

	case *Decrement:
		c.count--
		fmt.Printf("Decremented: %d\n", c.count)

	case *GetCount:
		ctx.Respond(&CountResponse{Count: c.count})

	case *Reset:
		c.count = 0
		fmt.Println("Reset to 0")

	default:
		fmt.Printf("Unknown message: %T\n", msg)
	}
}

// ============================================
// Main
// ============================================

func main() {
	fmt.Println("=== Counter Actor Example (Go) ===")
	fmt.Println()

	// Actor System 생성
	system := actor.NewActorSystem()

	// Counter Actor 생성
	props := actor.PropsFromProducer(func() actor.Actor {
		return &CounterActor{}
	})

	pid := system.Root.Spawn(props)

	// Tell 패턴: 응답 없이 메시지 전송
	system.Root.Send(pid, &Increment{})
	system.Root.Send(pid, &Increment{})
	system.Root.Send(pid, &Increment{})
	system.Root.Send(pid, &Decrement{})

	// 처리 대기
	time.Sleep(100 * time.Millisecond)

	// Ask 패턴: 응답 대기
	future := system.Root.RequestFuture(pid, &GetCount{}, 5*time.Second)
	result, err := future.Result()
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}

	response := result.(*CountResponse)
	fmt.Printf("\nFinal count: %d\n", response.Count)

	// Reset
	system.Root.Send(pid, &Reset{})
	time.Sleep(100 * time.Millisecond)

	future = system.Root.RequestFuture(pid, &GetCount{}, 5*time.Second)
	result, _ = future.Result()
	fmt.Printf("After reset: %d\n", result.(*CountResponse).Count)

	// Actor 종료
	system.Root.Stop(pid)
	time.Sleep(100 * time.Millisecond)

	// 동시성 테스트
	concurrencyDemo(system)
}

// ============================================
// Concurrency Demo
// ============================================

func concurrencyDemo(system *actor.ActorSystem) {
	fmt.Println("\n=== Concurrency Demo ===")
	fmt.Println()

	props := actor.PropsFromProducer(func() actor.Actor {
		return &CounterActor{}
	})

	pid := system.Root.Spawn(props)

	// 1000개의 동시 요청
	var wg sync.WaitGroup
	wg.Add(1000)

	for i := 0; i < 1000; i++ {
		go func() {
			defer wg.Done()
			system.Root.Send(pid, &Increment{})
		}()
	}

	wg.Wait()
	time.Sleep(500 * time.Millisecond)

	// 결과 확인
	future := system.Root.RequestFuture(pid, &GetCount{}, 5*time.Second)
	result, _ := future.Result()
	count := result.(*CountResponse).Count

	fmt.Printf("Expected: 1000, Actual: %d\n", count)
	if count == 1000 {
		fmt.Println("Thread-safe: ✅ YES")
	} else {
		fmt.Println("Thread-safe: ❌ NO")
	}

	system.Root.Stop(pid)
}

// ============================================
// 간단한 구현 (Proto.Actor 없이)
// ============================================

// 프레임워크 없이 개념 이해를 위한 구현

type SimpleMessage interface{}

type SimpleActor struct {
	mailbox chan SimpleMessage
	count   int
	done    chan struct{}
}

func NewSimpleActor() *SimpleActor {
	a := &SimpleActor{
		mailbox: make(chan SimpleMessage, 100),
		done:    make(chan struct{}),
	}
	go a.run()
	return a
}

func (a *SimpleActor) run() {
	for {
		select {
		case msg := <-a.mailbox:
			switch msg.(type) {
			case *Increment:
				a.count++
			case *Decrement:
				a.count--
			case *Reset:
				a.count = 0
			}
		case <-a.done:
			return
		}
	}
}

func (a *SimpleActor) Send(msg SimpleMessage) {
	a.mailbox <- msg
}

func (a *SimpleActor) Stop() {
	close(a.done)
}

File: examples/01-basic-actor/typescript/counter.ts
/**
 * Basic Actor Pattern - Counter Actor in TypeScript
 *
 * 이 예제는 Actor Model의 기본 개념을 직접 구현합니다.
 * 프로덕션에서는 전용 라이브러리 사용을 권장합니다.
 */

// ============================================
// Message Types (메시지 타입 정의)
// ============================================

type Message =
  | { type: 'INCREMENT' }
  | { type: 'DECREMENT' }
  | { type: 'GET_COUNT'; replyTo: (count: number) => void }
  | { type: 'RESET' };

// ============================================
// Actor Base Class (Actor 기본 클래스)
// ============================================

abstract class Actor<T> {
  private mailbox: T[] = [];
  private processing = false;

  // 메시지 전송 (비동기, 블로킹 없음)
  send(message: T): void {
    this.mailbox.push(message);
    this.processNext();
  }

  // 메시지 처리 (서브클래스에서 구현)
  protected abstract receive(message: T): void | Promise<void>;

  // Mailbox에서 메시지를 하나씩 처리
  private async processNext(): Promise<void> {
    if (this.processing) return;

    this.processing = true;

    while (this.mailbox.length > 0) {
      const message = this.mailbox.shift()!;
      try {
        await this.receive(message);
      } catch (error) {
        console.error('Error processing message:', error);
        // Actor 모델에서는 에러 시 Supervisor에게 알림
        // 여기서는 단순히 로그만 출력
      }
    }

    this.processing = false;
  }
}

// ============================================
// Counter Actor (카운터 액터 구현)
// ============================================

class CounterActor extends Actor<Message> {
  // 캡슐화된 상태 - 외부에서 직접 접근 불가
  private count: number = 0;

  protected receive(message: Message): void {
    switch (message.type) {
      case 'INCREMENT':
        this.count++;
        console.log(`Incremented: ${this.count}`);
        break;

      case 'DECREMENT':
        this.count--;
        console.log(`Decremented: ${this.count}`);
        break;

      case 'GET_COUNT':
        // 응답을 콜백으로 전달 (Ask 패턴)
        message.replyTo(this.count);
        break;

      case 'RESET':
        this.count = 0;
        console.log('Reset to 0');
        break;
    }
  }
}

// ============================================
// 사용 예제
// ============================================

async function main() {
  console.log('=== Counter Actor Example ===\n');

  const counter = new CounterActor();

  // Tell 패턴: 응답 없이 메시지 전송
  counter.send({ type: 'INCREMENT' });
  counter.send({ type: 'INCREMENT' });
  counter.send({ type: 'INCREMENT' });
  counter.send({ type: 'DECREMENT' });

  // Ask 패턴: 응답 대기
  const getCount = (): Promise<number> => {
    return new Promise((resolve) => {
      counter.send({ type: 'GET_COUNT', replyTo: resolve });
    });
  };

  // 비동기로 결과 확인
  await new Promise(resolve => setTimeout(resolve, 100));

  const count = await getCount();
  console.log(`\nFinal count: ${count}`);

  // Reset
  counter.send({ type: 'RESET' });

  await new Promise(resolve => setTimeout(resolve, 100));

  const resetCount = await getCount();
  console.log(`After reset: ${resetCount}`);
}

main().catch(console.error);

// ============================================
// 스레드 안전성 데모
// ============================================

async function concurrencyDemo() {
  console.log('\n=== Concurrency Demo ===\n');

  const counter = new CounterActor();

  // 동시에 1000개의 INCREMENT 메시지 전송
  const promises: Promise<void>[] = [];

  for (let i = 0; i < 1000; i++) {
    promises.push(
      new Promise((resolve) => {
        counter.send({ type: 'INCREMENT' });
        resolve();
      })
    );
  }

  await Promise.all(promises);

  // 잠시 대기 후 결과 확인
  await new Promise(resolve => setTimeout(resolve, 500));

  const getCount = (): Promise<number> => {
    return new Promise((resolve) => {
      counter.send({ type: 'GET_COUNT', replyTo: resolve });
    });
  };

  const finalCount = await getCount();
  console.log(`Expected: 1000, Actual: ${finalCount}`);
  console.log(`Thread-safe: ${finalCount === 1000 ? '✅ YES' : '❌ NO'}`);
}

// concurrencyDemo().catch(console.error);
